### PR TITLE
Feature/camera x

### PR DESCRIPTION
**Pull request by apalala-dev**
## Description

This PR intends to migrate the app to CameraX among other improvements to the Android platform, including:
- Rewrite Android API to use CameraX instead of Camera2
- Start Kotlin migration
- Use pigeon instead of MethodChannel when possible on Android
- Add Video capture on Android (with pause and resume)
- Add ability to save GPS location in Exif for pictures

## Checklist

Before creating any Pull Request, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).

- [x] 📕 I read the [Contributing page](https://github.com/Apparence-io/camera_awesome/blob/master/CONTRIBUTING.md).
- [x] 🤝 I match the actual coding style.
- [x] ✅ I ran ```flutter analyze``` without any issues.

## Breaking Change

- [ ] 🛠 My feature contain breaking change.

This PR brings a lot of changes and must be heavily tested. There should not be breaking changes, but since the CameraX API is quite different from Camera2, there might be some differences.
iOS could not be tested as I don't have a physical iOS device.

### CODE REVIEW

**Comment by apalala-dev** [comment]
## GPS Position
- Should we capture the GPS position before taking the picture (as done now)? It adds some delay to the actual image capture

## ImageAnalysis
- ImageAnalysis resolution is hardcoded. I'd like to give users the option to choose it, but I am not sure if this is available on iPhone. Any idea @dim-apparence ?
- An example using the ImageAnalysis could be nice (decoding barcodes?)
- ImageAnalysis does not rotate the picture (but since we get the orientation on the client side, it could be up to them to do it). It seemed to not be rotated on master.

## Sized preview
- Camera preview size seems to be stretched to the screen size. On my screen for instance, the higher camera resolution I get is 1600x1200. I asked for a resolution of 3000x3000 (and pictures taken are with this resolution). This implies that the preview I get in response is of a different aspect ratio than what I asked.
- Sometimes the transition between full screen and sized preview does not work very well in the example app.

## Focus
- Focus on a specific point is not that hard to do with CameraX. The code to do it is included but the UX part is not done. iOS is not done neither.

File: android/src/main/kotlin/com/apparence/camerawesome/CameraPreviewTest.kt [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+//            ), ArgumentMatchers.eq(CameraMetadata.FLASH_MODE_SINGLE)
+//        )
+//        cameraPreview!!.setFlashMode(FlashMode.ALWAYS)
+//        Mockito.verify(captureRequestBuilder, Mockito.atLeastOnce()).set(
+//            ArgumentMatchers.eq(
+//                CaptureRequest.FLASH_MODE
+//            ), ArgumentMatchers.eq(CameraMetadata.FLASH_MODE_TORCH)
+//        )
+//    }
+//}
```
Comment (g-apparence) [inline android/src/main/kotlin/com/apparence/camerawesome/CameraPreviewTest.kt:123]: Si plus util => delete

File: android/src/main/kotlin/com/apparence/camerawesome/cameraX/CameraXState.kt [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+    @SuppressLint("RestrictedApi", "UnsafeOptInUsageError")
+    fun updateLifecycle(activity: Activity) {
+        // Preview
+        preview = Preview.Builder().setCameraSelector(cameraSelector).build()
+        preview!!.setSurfaceProvider(
+            surfaceProvider(executor(activity))
+        )
+
+        if (currentCaptureMode == CaptureModes.PHOTO) {
```
Comment (g-apparence) [inline android/src/main/kotlin/com/apparence/camerawesome/cameraX/CameraXState.kt:66]: Split en photo / video state avec une super class ?

File: android/src/main/kotlin/com/apparence/camerawesome/cameraX/CameraXState.kt [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+                .build()
+        } else {
+            recorder = Recorder.Builder()
+                .setQualitySelector(QualitySelector.from(Quality.HIGHEST))
+                .build()
+            videoCapture = VideoCapture.withOutput(recorder!!)
+        }
+        var imageAnalysis: ImageAnalysis? = null
+        if (enableImageStream) {
+            // TODO implement imageanalysis usecase
```
Comment (g-apparence) [inline android/src/main/kotlin/com/apparence/camerawesome/cameraX/CameraXState.kt:88]: remove todo

File: android/src/main/kotlin/com/apparence/camerawesome/cameraX/CameraXState.kt [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+                // TODO What should the targetResolutionSize be?
+                .setTargetResolution(Size(640, 480))
+                // TODO Should backpressure be a parameter?
+                .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
+                .setOutputImageFormat(ImageAnalysis.OUTPUT_IMAGE_FORMAT_YUV_420_888)
+                .build()
+            imageAnalysis.setAnalyzer(executor(activity)) { imageProxy ->
+                if (previewStreamSink != null) {
+                    // TODO Not sure of the benefits of running the conversion in the background
+                    // Copying data between threads might be expensive
```
Comment (g-apparence) [inline android/src/main/kotlin/com/apparence/camerawesome/cameraX/CameraXState.kt:99]: Du coup niveau perf ca passe aujourd'hui comme ça ?

File: lib/camerapreview.dart [hunk 5]
```diff
@@ -210,15 +214,15 @@ class CameraAwesomeState extends State<CameraAwesome>
...
-      });
+      // _permissionStreamSub =
+      //     CamerawesomePlugin.listenPermissionResult()!.listen((res) {
+      //   if (res) {
+      //     initPlatformState();
+      //   }
+      //   if (widget.onPermissionsResult != null) {
+      //     widget.onPermissionsResult!(res);
+      //   }
+      // });
```
Comment (g-apparence) [inline lib/camerapreview.dart:225]: Du coup c'est plus un stream si je comprends bien ?

**Review by g-apparence — COMMENTED** [review]
C'est effectivement un gros pack. 
Je vais aussi faire plus de tests pour check les bugs. 

Tu peux attacher la branche au repo pour que on puisse aller dessus ?

File: lib/camerapreview.dart (hunk 5 above)
Comment (apalala-dev) [reply lib/camerapreview.dart]: C'est un oubli, ça devrait être corrigé dans un de mes derniers commits (j'ai rétabli le fonctionnement d'avant)

File: android/src/main/kotlin/com/apparence/camerawesome/cameraX/CameraXState.kt (hunk 2 above)
Comment (apalala-dev) [reply android/src/main/kotlin/com/apparence/camerawesome/cameraX/CameraXState.kt:66]: Pas sûr que ça soit pertinent car on peut potentiellement avoir de la capture vidéo ET photo en même temps sur CameraX (plusieurs use cases en même temps). C'est pas permis par CamerAwesome pour le moment, mais ça pourrait l'être à l'avenir ?

File: android/src/main/kotlin/com/apparence/camerawesome/CameraPreviewTest.kt (hunk 1 above)
Comment (apalala-dev) [reply android/src/main/kotlin/com/apparence/camerawesome/CameraPreviewTest.kt:123]: Je me suis encore penché sur les tests. Il faudra me donner un coup de main pour m'expliquer ce qui est en place sur le projet et que je puisse vérifier que ça marche toujours / compléter au besoin

**Pull request merged**